cpp/src/toppra/constraint.cpp

```diff
@@ -0,0 +1,43 @@
+#include <toppra/constraint.hpp>
+
+namespace toppra {
+
+std::ostream& LinearConstraint::print(std::ostream& os) const
+{
+  os << "LinearConstraint\n"
+        "    Type: " << constraintType_ << "\n"
+        "    Discretization Scheme: " << discretizationType_ << "\n";
+  return os;
+}
+
+void LinearConstraint::discretizationType (DiscretizationType type)
+{
+  // I don't think the check done in Python is useful in C++.
+  discretizationType_ = type;
+}
+
+void LinearConstraint::computeParams(const GeometricPath& path, const Vector& gridpoints,
+    Matrix& a, Matrix& b, Matrix& c, Matrix& F, Matrix& g, Bounds& ubound, Bounds& xbound)
+{
+  Eigen::Index N_1 = gridpoints.size(),
+               m = a.cols(),
+               k = g.cols();
+  if (b.cols() != m)
+    std::invalid_argument("Wrong number of columns of matrix b");
+  if (c.cols() != m)
+    std::invalid_argument("Wrong number of columns of matrix c");
+  //if (F.cols() != m)
```

Comment:
We will need F to represent the constraint. Why is it commented out?

Reply:
Because F is not of the good type yet. It is a tensor. So either we rely on the Eigen Tensor module or we use an internal type.
At the moment, the constraint compute the parameters of all the gridpoints at once. I wonder if it would make sense to compute the parameter for on point only.

The pros are:
- this would make it easier to handle constraints only on a subinterval,
- we would not need tensors.

Reply:
Ah, right I miss that point. Actually we do not use the fact that F is a tensor, so I think it's find letting F be a vector of matrices.

On a related note, probably it is more reasonable to have a,b,c, g be vectors of 1d-matrices as well.

Reply:
Ok. One question while we are at on this point. I have scenarios where it would make sense to have certain constraints only on a sub-interval. Do you think we can allow this ?

Reply:
From the algorithmic viewpoint, it's perfectly fine. You only need to set F[i] for i outside of the interested range to be zero, and g[i] to -1. 

From a physical perspective, however, the discontinuity can be worrying. You would probably want to slowly relax the constraint instead of abruptly set the coefficients to these values.

Reply:
Won't the abrupt change of constraint be smoothed by the forward and backward pass ? If the constraint you add is a limit of the cartesian velocity of the end-effector, the end-effector velocity should *smoothly* decrease a few step ahead the constraint change.

Reply:
Yes, mathematically it will work. The robot will slow down just nice to satisfy other constraints. 

However, you never until you try the code on a real, physical robot.